### ISSUE TITLE
add python to the documentation

### ISSUE THREAD

**Issue by yoavg** [clab/dynet#169]
Currently, the documentation is about the c++ modules.

Would be nice to be explicit about this, and also add links to some of the python resources.
While these are not yet very nice or exhaustive, we do have some useful resources:

https://github.com/clab/dynet/blob/master/python/CHANGES.md
https://github.com/clab/dynet/tree/master/pyexamples
https://github.com/clab/dynet/tree/master/pyexamples/tutorials

and of course the EMNLP tutorial (with examples and slides):
https://github.com/clab/dynet_tutorial_examples

These should be added to the main documentation page.
I'd do it myself, but I am not sure on how the documentation system is set up.
@pmichel31415 ?

**Comment by pmichel31415** [comment]
So there are detailed guidelines in the documentation, under the "contributing" section [here](http://dynet.readthedocs.io/en/latest/doc_style.html)

TL;DR

If you want to write a tutorial you can directly modify the documentation source in [doc/source](https://github.com/clab/dynet/tree/master/doc/source) (careful, this uses rst markup, but you can convert from markdown using pandoc).

If you want to write a reference manual, sphinx can retrieve the documentation from python docstrings directly, although I am not certain it is straightforward with cython ([this stackoverflow question](http://stackoverflow.com/questions/10057476/how-do-i-use-sphinx-with-cython) might help).

For ipython notebooks I don't know if they can be embedded directly, but you can probably link them from the docs.

**Comment by pmichel31415** [comment]
OK I came back to this because this is the biggest barrier of entry right now IMO.

Here's the state of things : 

1. Generating the doc automatically with sphinx (like what we do for c++) requires to compile the cython files (sphinx doesn't analyse the source code apparently, don't ask me why)
2. This means if we want to automate things completely on readthedocs like c++ we need to recompile the whole of dynet on readthedocs before generating the docs (impractical)

I see 2 solutions to this :

1. Cython should be documented with standard python docstrings (this allows at least to call help in the python/ipython interpreter)
2. A(n ugly) workaround for sphinx would be to have a manually compiled `_dynet.so` file on the repo and try to fetch the docs directly from this.

I can start working directly on 1. however I need your opinion on 2. @yoavg @neubig

**Comment by neubig** [comment]
One thing I was thinking about was that it might be worth totally assisting the doc for Python and C++. in other words, the would be totally different readthedocs sites. Would this make things easier?

**Comment by pmichel31415** [comment]
Hmm, I don't think this would solve the main problem : we need to recompile dynet on readthedocs.

Also I'm not sure readthedocs allows more than one site for one repo (although we can play with branches)

**Comment by pmichel31415** [comment]
TODO : 

- [x] DynetParams
- [x] Parameter
- [ ] LookupParameter
- [ ] Saveable
- [ ] PyInitializer(s)
- [ ] Model
- [ ] ComputationGraph (? : should'nt be used by the user but could be useful for contributors) 
- [x] Expression(s)
- [ ] Operations
- [ ] RNNBuilder(s)
- [ ] RNNState
- [ ] Trainers
- [ ] misc functions (renew_cg, etc...)

**Comment by neubig** [comment]
Sorry, getting back to the original question of how to get this to work on readthedocs, I think Chainer is using readthedocs, and their cupy library uses Cython. Would taking a look at this help at all?
https://github.com/pfnet/chainer/tree/master/docs

**Comment by pmichel31415** [comment]
OK this is also mostly done, I'll close this one
